src/box/box.cc

```diff
@@ -797,14 +823,136 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
 	return -1;
 }
 
+/* Storage, adding rows to a rlist to separate mixed transactions. */
+struct wal_row {
+	/* Base class. */
+	struct xrow_header row;
+	/* A link on the list of rows stacked in the nodes_rows array. */
+	struct rlist in_row_list;
+	/* A growing identifier to track lsregion allocations. */
+	int64_t lsr_id;
+};
+
+/*
+ * Callback to stash row and row bodies upon receipt, used to recover
+ * mixed transactions.
+ */
+static struct wal_row *
+wal_stream_save_row(struct wal_stream *stream, struct xrow_header *row)
+{
+	static int64_t lsr_id = 0;
+	struct lsregion *lsr = &stream->lsr;
+	struct wal_row *new_row =
+		(struct wal_row *)xlsregion_alloc(lsr, sizeof(struct wal_row),
+						  ++lsr_id);
+	new_row->lsr_id = lsr_id;
+	memcpy(new_row, row, sizeof(struct xrow_header));
```

Comment:
You can just do `new_row->row = *row`. Doing `memcpy()` over different struct types should be better avoided. It will break if you add any member before `wal_row::row`.

Reply:
Done.